File: .net/dojos/dojo1/Bowling/Bowling/Frame.cs
﻿namespace Bowling
{
    public class Frame
    {
        public int FirstTry { get; set; }
        public int SecondTry { get; set; }

        public int Score
        {
            get { return CalculateScore(); }
        }


        public Frame Next { get; set; }

        private int CalculateScore()
        {
            int score= FirstTry + SecondTry;
            if (IsStrike())
            {
                score += (NextBall()+NextToNextBall());
            }
            else if(IsSpare())
            {
                score += NextBall();
            }
            return score;
        }

        private int NextToNextBall()
        {
            if (Next.IsStrike())
            {
               return Next.Next.FirstTry;
            }
            return Next.SecondTry;
        }

        private int NextBall()
        {
            return Next.FirstTry;
        }

        private bool IsSpare()
        {
            return FirstTry+SecondTry==10;
        }

        private bool IsStrike()
        {
            return FirstTry == 10;
        }
    }
}

File: dotnet/dojos/dojo1/JBG/src/main/java/bowling/second/practice/Frame.java
package bowling.second.practice;

public class Frame {
    private int firstBall;
    private int secondBall;
    private int ownScore;
    private Frame nextFrame;

    public Frame(int firstBall, int secondBall) {
        this.firstBall = firstBall;
        this.secondBall = secondBall;
        this.ownScore = firstBall + secondBall;
    }

    public int countScore() {
        return ownScore + countBonus();
    }

    public void setNextFrame(Frame nextFrame) throws Exception {
        this.nextFrame = nextFrame;
    }

    private int countBonus() {
        int bonus = 0;
        if (isStrike()) {
            bonus = nextBall() + nextNextBall();
        } else if (isSpare()) {
            bonus = nextBall();
        }
        return bonus;
    }

    private boolean isSpare() {
        return ownScore == 10;
    }

    private boolean isStrike() {
        return firstBall == 10;
    }

    private int nextNextBall() {
        if (nextFrame.isStrike() && nextFrame.isNotLast()) {
            return nextFrame.nextBall();
        }
        return nextFrame.secondBall;
    }

    private boolean isNotLast() {
        return nextFrame != null;
    }

    private int nextBall() {
        return nextFrame.firstBall;
    }
}
